### PR TITLE
Add operator benchmarking run to CI nightly

### PR DESCRIPTION
**Pull request by jainapurva**
This PR introduces a new "operator microbenchmark" CI workflow and GitHub Actions for operator microbenchmarks, updating test scripts and job matrices to support new parameters, and broadening the operator benchmark tests to include more data types, larger shapes, and gradient tests. The benchmark configurations now focus more on different cuda hardware and multiple dtypes (bf16, fp16, fp32), for both compile and eager mode.

**Benchmark Configuration and Coverage:**

* Expanded operator benchmark configurations in `addmm_test.py`, `bmm_test.py`, `matmul_test.py`, and `mm_test.py` to benchmark multiple dtypes on CUDA devices, in eager and compile mode, for forward and backward run. The configs with tag "long" for the above mentioned files are being run in CI.
* The CI benchmarking is running on various hardwares: H100, A100.
* The CI job also uploads the microbenchmarking outputs to a [HUD](https://hud.pytorch.org/benchmark/llms?repoName=pytorch%2Fpytorch&benchmarkName=PyTorch+operator+microbenchmark) dashboard.

### CODE REVIEW

File: .github/workflows/operator_microbenchmark.yml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      build-environment: linux-jammy-cuda12.8-py3.10-gcc9-sm80
+      docker-image-name: ci-image:pytorch-linux-jammy-cuda12.8-cudnn9-py3-gcc11
+      cuda-arch-list: '8.0 9.0'
+      test-matrix: |
+        { include: [
+          { config: "operator_microbenchmark_test",
+            shard: 1,
+            num_shards: 1,
+            runner: "linux.aws.h100",
+            op_benchmark_tests: [matmul, mm, add, bmm],
```
Comment (huydhn) [inline .github/workflows/operator_microbenchmark.yml]: I don't think GH like this syntax https://github.com/pytorch/pytorch/actions/runs/17687955792/job/50277726408.  So we have two choices:

1. Let's put this logic in .ci/pytorch/test.sh.  You could choose a suitable name for the config, which will be passed into the script.  That's the usual approach.
2. You can try make them into a JSON strings, for example, the parse the strings in .ci/pytorch/test.sh, or pass them string to op benchmark python script to handle.
```
op_benchmark_tests: '["matmul", "mm", "add", "bmm"]',
extra_flags: "['--use-compile', '']"
```

File: .github/workflows/operator_microbenchmark.yml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+  opmicrobenchmark-test:
+    name: opmicrobenchmark-test
+    uses: ./.github/workflows/_linux-test.yml
+    needs: opmicrobenchmark-build
+    with:
+      timeout-minutes: 500
+      build-environment: linux-jammy-cuda12.8-py3.10-gcc9-sm80
+      docker-image: ${{ needs.opmicrobenchmark-build.outputs.docker-image }}
+      test-matrix: ${{ needs.opmicrobenchmark-build.outputs.test-matrix }}
```
Comment (huydhn) [inline .github/workflows/operator_microbenchmark.yml:45]: To be able to run this on B200, you need this additional line https://github.com/pytorch/pytorch/blob/main/.github/workflows/test-b200.yml#L75 to tell the runner to use this role to login and grab the Docker image `ci-image:pytorch-linux-jammy-cuda12.8-cudnn9-py3-gcc11`

```suggestion
      test-matrix: ${{ needs.opmicrobenchmark-build.outputs.test-matrix }}
      aws-role-to-assume: arn:aws:iam::308535385114:role/gha_workflow_s3_and_ecr_read_only
```

**Review by huydhn — APPROVED** [review]
LGTM!

File: .github/workflows/_linux-test.yml [hunk 3]
```diff
@@ -123,8 +123,8 @@ jobs:
         uses: ./.github/actions/setup-linux
         if: inputs.build-environment != 'linux-s390x-binary-manywheel' && !contains(matrix.runner, 'b200')
 
-      - name: configure aws credentials
-        if: ${{ inputs.aws-role-to-assume != '' && inputs.build-environment != 'linux-s390x-binary-manywheel' }}
+      - name: Configure aws credentials
```
Comment (huydhn) [inline .github/workflows/_linux-test.yml]: Just FYI, this is a minor tweak to make the workflow working with both A100/H100 and B200

File: .ci/pytorch/test.sh [hunk 4]
```diff
@@ -1614,6 +1614,25 @@ test_operator_benchmark() {
...
+  TEST_REPORTS_DIR=$(pwd)/test/test-reports
+  mkdir -p "$TEST_REPORTS_DIR"
+  TEST_DIR=$(pwd)
+
+  cd benchmarks/operator_benchmark/pt_extension
+  python -m pip install .
+
+  cd "${TEST_DIR}"/benchmarks/operator_benchmark
+
+  for OP_BENCHMARK_TESTS in matmul mm addmm bmm; do
```
Comment (jbschlosser) [inline .ci/pytorch/test.sh:1643]: nit: can you factor these out somewhere (e.g  in a list or something) for ease of maintenance?

File: benchmarks/operator_benchmark/pt/matmul_test.py [hunk 5]
```diff
@@ -13,33 +13,46 @@
...
+                M, N, device=device, dtype=dtype, requires_grad=self.auto_set()
+            )
             if trans_a
-            else torch.rand(N, M, device=device).t(),
-            "input_two": torch.rand(N, K, device=device)
+            else torch.rand(N, M, device=device, dtype=dtype)
+            .t()
+            .contiguous()
+            .clone()
+            .detach()
```
Comment (jbschlosser) [inline benchmarks/operator_benchmark/pt/matmul_test.py]: what's the reason behind these calls?

File: benchmarks/operator_benchmark/pt/bmm_test.py [hunk 6]
```diff
@@ -27,21 +27,25 @@
 )
 
 batched_binary_configs_long = op_bench.cross_product_configs(
-    B=[1, 128],
-    M=[8, 128],
-    N=[32, 64],
-    K=[4, 256],
-    device=["cpu", "cuda"],
-    dtype=[torch.float, torch.bfloat16],
+    B=[8, 32],
```
Comment (jbschlosser) [inline benchmarks/operator_benchmark/pt/bmm_test.py:30]: I'm good with the updated benchmark configs throughout, but I'm curious what prompted the changes?

**Review by jbschlosser — COMMENTED** [review]
it'll be awesome to have these running in CI, thanks :)

File: benchmarks/operator_benchmark/pt/bmm_test.py (hunk 6 above)
Comment (jainapurva) [reply benchmarks/operator_benchmark/pt/bmm_test.py:30]: I added some larger shapes, and more dtypes for an extensive benchmarking, with real world shapes. There's a separate operator benchmarking already running for [cpu](https://github.com/pytorch/pytorch/actions/workflows/operator_benchmark.yml), hence to reduce the load on GPUs I removed cpu from the long tag benchmarks, which I'm using for benchmarking on cuda devices.

File: benchmarks/operator_benchmark/pt/matmul_test.py (hunk 5 above)
Comment (jainapurva) [reply benchmarks/operator_benchmark/pt/matmul_test.py]: Removed them in cleanup

File: .ci/pytorch/test.sh (hunk 4 above)
Comment (jainapurva) [reply .ci/pytorch/test.sh:1643]: I had tried making it a separate list, but with the current process of reading CI files, it would involve a wider update in the CI workflows, hence I believe that should be separated as an additional improvement for future
